### PR TITLE
Add -package-name to driver

### PR DESCRIPTION
**Pull request by elsh**
Add -package-name to driver
Add packageName to ModuleOutputInfo
Add package name validation and diagnostics
If package name is empty or invalid, fall back to module name
Resolves rdar://105119573.

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by artemcm** [comment]
@elsh could you please update `Options.swift` with these instructions:
https://github.com/apple/swift-driver#rebuilding-optionsswift ?

This should bring in `-package-name` and any other newly added/modified flags.

**Comment by artemcm** [comment]
Oh and this repo does not have a smoke test, just `@swift-ci test`.

**Comment by elsh** [comment]
@swift-ci test

**Review by artemcm — APPROVED** [review]

**Review by nkcsgexi — APPROVED** [review]

**Comment by elsh** [comment]
@swift-ci test macOS

**Comment by elsh** [comment]
@swift-ci test macOS

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test macOS

**Comment by elsh** [comment]
@swift-ci test

File: Sources/SwiftDriver/Driver/Driver.swift [hunk 1]
```diff
@@ -2427,9 +2427,30 @@ extension Driver {
     // Retrieve and validate module aliases if passed in
     let moduleAliases = moduleAliasesFromInput(parsedOptions.arguments(for: [.moduleAlias]), with: moduleName, onError: diagnosticsEngine)
 
+    // Validate package name and fall back to module name or module alias if it exists
+    var fallback = true
```
Comment (artemcm) [inline Sources/SwiftDriver/Driver/Driver.swift:2431]: Nit: Could you please make this variable describe which fallback it's referring to?

**Review by artemcm — APPROVED** [review]
LGTM!

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 2]
```diff
@@ -689,6 +689,28 @@ final class SwiftDriverTests: XCTestCase {
...
+    try assertNoDriverDiagnostics(args: "swiftc", "file.swift", "bar.swift", "-module-name", "MyModule", "-package-name", "mypkg", "-emit-module", "-emit-module-path", "../../path/to/MyModule.swiftmodule") { driver in
+        XCTAssertEqual(driver.moduleOutputInfo.packageName,  "mypkg")
+        XCTAssertEqual(driver.moduleOutputInfo.output, .topLevel(try VirtualPath.intern(path: "../../path/to/MyModule.swiftmodule")))
+    }
+  }
+
+  func testPackageNameFallbacks() throws {
+    try assertDriverDiagnostics(args: ["swiftc", "file.swift"]) { // -package-name is nil
+      $1.expect(.warning("package name is nil; will fall back to module name \"file\""))
+    }
```
Comment (xymus) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:702]: I don't think the warning on no package name is necessary. If we write a custom driver command it's usually for tests or simple scripts, we may just not care about package visibility in that case and we don't want a warning at that point.

If no package is defined I would just make it a warning or error in the code that uses that `package` access level. So if you don't use `package` and don't define a package name then you won't see any diagnostics.

**Review by xymus — APPROVED** [review]
I think it's worth removing the warning when not using package at all. Otherwise this looks good!

File: Tests/SwiftDriverTests/SwiftDriverTests.swift (hunk 2 above)
Comment (artemcm) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:702]: Oh! Agreed with @xymus, we should not warn here.

File: Sources/SwiftDriver/Driver/Driver.swift (hunk 1 above)
Comment (elsh) [reply Sources/SwiftDriver/Driver/Driver.swift]: no more fallback; now emits an error instead

File: Tests/SwiftDriverTests/SwiftDriverTests.swift (hunk 2 above)
Comment (elsh) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:703]: updated

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test macOS

**Pull request merged**